Model Details and Model Version Details: add editable properties section, set up API patches for description, labels and properties

mturley opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->

Closes [RHOAIENG-2235](https://issues.redhat.com/browse/RHOAIENG-2235) and [RHOAIENG-2238](https://issues.redhat.com/browse/RHOAIENG-2238)
Followup to https://github.com/opendatahub-io/odh-dashboard/pull/2760

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

* Implements the contents of the Properties section using a new `ModelPropertiesDescriptionListGroup` component which provides view/add/edit/delete controls
* Implements API patch requests when saving edits to Description, Labels and Properties (replacing the stub Promise for the former two)
* Adds input validation to ensure labels and property keys are unique. See **"Input validation screencaps"** below.
* None of these components/utils are specific to `RegisteredModel`, they work with `ModelRegistryCustomProperties` objects and generic `T extends ModelRegistryBase` objects, so it should be easy to reuse these for the ModelVersion detail page.
* Note that the patch requests must use the new `getPatchBody` utility in order to make sure the request isn't clearing other fields due to the way patches are performed in the API. We must supply all editable fields when patching an object even if we aren't modifying them all, but we can't supply some fields like the create/edit timestamps or we'll get 400 errors.

### Demo screencaps

* Editing the Description of a RegisteredModel

  https://github.com/opendatahub-io/odh-dashboard/assets/811963/08721aae-5897-4fa3-a00e-a7e7bee78ecb

* Editing the Labels of a RegisteredModel

  https://github.com/opendatahub-io/odh-dashboard/assets/811963/6bed2da9-99cc-4880-942b-e5e981cc4609

* Editing the Properties of a RegisteredModel

  https://github.com/opendatahub-io/odh-dashboard/assets/811963/4aba24ad-f97e-42b5-a49b-4be3432fc6cf

### Input validation screencaps
* When adding a new label, in addition to the existing validation that the label text must be no longer than 63 characters, we also now validate that a newly added label is unique (does not match any existing key in `customProperties`, whether that is a label or the key of a property in the table below). If the label is not unique, the modal submission is blocked and an error message appears:
  
  <img width="584" alt="Screenshot 2024-05-02 at 3 44 56 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/5ebb0b79-fce6-4a57-ae02-9c456a9a58d5">
    
* When editing an existing label, the same validation is applied. However, with the [current beta version of PF editable labels](https://www.patternfly.org/components/label#editable-labels), there is no way to express input validation errors while editing a label's text. As a workaround for now, if the user attempts to use an invalid label (either longer than 63 characters or matching an existing label/property), when they click away the change is not applied and the label remains as it was before they clicked it:
  
  https://github.com/opendatahub-io/odh-dashboard/assets/811963/d88261bf-0743-4f36-9c41-2fe9a43964cc

* When adding or editing a property, the same validation is applied to the property key. Errors in this case appear inline below the key input field:
  
  <img width="725" alt="Screenshot 2024-05-02 at 3 53 40 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/16c2b375-db95-49a1-abb3-fa0766404cf6">
  <img width="749" alt="Screenshot 2024-05-02 at 3 52 26 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/d25af2c4-8da6-423f-a3e8-181dbfca95f3">

cc @yih-wang @vconzola 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Go to `Model Registry` in the nav, click a registered model and click the Details tab. Play around with the edit controls on the Description, Labels and Properties sections. (see the screen recordings above)

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->

Unit tests have been added for the new utility functions `mergeUpdatedLabels`, `getProperties`, `mergeUpdatedProperty` and `getPatchBody`. Cypress tests for all these views are to come soon.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

gitdallas commented on frontend/src/pages/modelRegistry/screens/ModelPropertiesTableRow.tsx:
```diff
@@ -0,0 +1,176 @@
+  const keyValidationError =
+    unsavedKey !== key && allExistingKeys.includes(unsavedKey)
+      ? 'Key must not match an existing property key or label'
+      : unsavedKey.length > 63
```
nit.. wonder if it would be better to just use normal if/else blocks here just for readability

mturley replied:
Good call

manaswinidas commented on frontend/src/pages/modelRegistry/screens/ModelPropertiesDescriptionListGroup.tsx:
```diff
@@ -0,0 +1,108 @@
+  customProperties = {},
+  saveEditedCustomProperties,
+}) => {
+  const [editingPropertyKeys, setEditingPropertyKeys] = React.useState<string[]>([]);
```
I'm not seeing the "Show more" functionality in Properties section. It's breaking the layout.

<img width="997" alt="Screenshot 2024-05-03 at 6 32 58 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/22885912/e3e17a41-973a-4b13-bc30-bba554e097e0">

mturley replied:
Resolved, see screenshots in comment below

manaswinidas reviewed (dismissed):
Also, I cannot see the dragging functionality for description. As per the mocks
<img width="248" alt="Screenshot 2024-05-03 at 6 39 45 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/22885912/4896fcfc-def1-4463-ba49-c2920c666cfa">

vconzola commented:
LGTM. @mturley I think the way you're handling error cases is fine for now. I also like the enhancement you suggested in Slack. I assume that'll be in a future PR once UXD has a chance to review your proposal.

mturley commented:
@manaswinidas I'm working on the overflow in the Properties section, but for the Description textarea I do get a draggable control in the bottom-right on Chrome. Although it should probably default to that 24-line height, I can change that.

mturley commented:
@manaswinidas I addressed your feedback: the description textarea will now default to 24 lines and be resizable only vertically (I was missing the `resizeOrientation` prop), and the Properties section now properly wraps and truncates text:

<img width="742" alt="Screenshot 2024-05-03 at 11 41 44 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/6274d712-d11b-4609-9779-34d15b6848a4">

<img width="754" alt="Screenshot 2024-05-03 at 11 41 50 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/5b0afaed-dbd4-4ab0-ba5b-cb367b7396a9">

<img width="742" alt="Screenshot 2024-05-03 at 11 42 01 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/811963/32bf6995-b4c3-401a-a482-394f503395a6">

Really good catch there, thank you!

mturley commented:
/hold
we'll merge https://github.com/opendatahub-io/odh-dashboard/pull/2774 first

mturley commented on frontend/src/components/EditableLabelsDescriptionListGroup.tsx:
```diff
@@ -107,7 +122,11 @@ const EditableLabelsDescriptionListGroup: React.FC<EditableTextDescriptionListGr
                 closeBtnProps={{ isDisabled: isSavingEdits }}
-                onEditComplete={(_event, newText) => editUnsavedLabel(newText, index)}
+                onEditComplete={(_event, newText) => {
+                  if (!reservedKeys.includes(newText) && newText.length <= 63) {
```
Yeah, I'll keep the discussion open about this and open an issue if we decide to change it. ([slack thread here](https://redhat-internal.slack.com/archives/C06EJG57Z0X/p1714681683554989))

mturley commented on frontend/src/concepts/modelRegistry/types.ts:
```diff
@@ -35,14 +35,65 @@ export enum ModelArtifactState {
+
+export type ModelRegistryCustomPropertyInt = {
+  metadataType: ModelRegistryMetadataType.INT;
+  int_value: string; // int64-formatted string
```
This is the way the data is structured in the API, I was going from the [swagger docs here](https://editor.swagger.io/?url=https://raw.githubusercontent.com/opendatahub-io/model-registry/main/api/openapi/model-registry.yaml). We could bring that feedback up with the model-registry backend team maybe, although I'm not sure we'll ever use the non-string types here. I just added them for completeness

yih-wang commented:
Hi @mturley, thanks for all the work, they look really great! I got two comments regarding the current screens, but feel free to close this pr and address the comments in future enhancements.

* It seems that the validation for duplicated labels/properties occurs while the user is typing. This could be annoying if the user hasn't finished inputting and accidentally duplicates an existing label. I believe it would be better to validate this after the user submits the modal/row.
* I didn't see from the screen cap how it works if the properties are more than 5 and collapsed, and the user clicks 'add property' at this moment. Just to clarify, the ideal behavior would be for the collapsed list to expand, and the cursor to automatically locate to the newly added key field.

mturley commented:
Hi @yih-wang,

Thanks for the feedback! Let's make sure to follow up with a UX issue / discussion to improve this after this PR.
* I'm not sure how input validation works elsewhere in the dashboard, I think it would be important to be consistent with that. I implemented the real-time validation because it is simpler to implement that way in React and I've seen it done that way elsewhere, but we can certainly change that behavior. This is one area where I may need to keep my non-designer opinion to myself, but personally as a user I always hate when I don't see validation until after I submit something because it takes me longer to try multiple times than to get the proactive feedback as I type and make sure I'm doing it right the first time. But I think I may be the weird one here!
* Currently if you click "Add property" while the properties are collapsed, it puts the new row at the bottom of the currently shown rows without expanding the rest. That's something we could easily change.

lucferbux commented on frontend/src/pages/modelRegistry/screens/ModelDetailsView.tsx:
```diff
@@ -1,97 +1,103 @@
+                  getPatchBodyForRegisteredModel(rm, { description: value }),
+                  rm.id,
+                )
+                .then(refreshAPIState)
```
Why do you want to call refresh api here? This just reload the fetching of the CR, which is not needed when PATCH right?

We should refresh ONLY the version details here, we are reloading pretty much everything.

lucferbux replied:
For this you can get the `refresh` value form `useModelVersionById` that we get in `ModelVersionDetails` and pass it to the component.

lucferbux replied:
I've just created a commit (cause i wanted to test it) which is more self-explanatory, I'll leave it here so you can check the suggestion: https://github.com/lucferbux/odh-dashboard/commit/6b921040d92a6d6cb0b366e8ac81a1fcb65051d1

mturley replied:
Ah, good call thank you! I was looking at `apiState` for a way to do that, of course it makes sense that it would be on `useModelVersionById`

lucferbux commented on frontend/src/pages/modelRegistry/screens/ModelDetailsView.tsx:
```diff
@@ -1,97 +1,103 @@
+                  }),
+                  rm.id,
+                )
+                .then(refreshAPIState)
```
Same here

lucferbux commented on frontend/src/pages/modelRegistry/screens/ModelVersionDetails/ModelVersionDetailsView.tsx:
```diff
@@ -1,20 +1,26 @@
 
 const ModelVersionDetailsView: React.FC<ModelVersionDetailsViewProps> = ({ modelVersion: mv }) => {
   const [modelArtifact] = useModelArtifactsByVersionId(mv.id);
+  const { apiState, refreshAPIState } = React.useContext(ModelRegistryContext);
```
Same here

lucferbux reviewed (changes requested):
Ok, so far everything seems great, I think there was still a discussion open right? But so far it looks great, just a small nit about refreshing data, we should use the right refreshing state for each component

lucferbux reviewed (approved):
/lgtm
